test/e2e/storage/testsuites/volumeperf.go

```diff
@@ -0,0 +1,287 @@
+			}
+		}()
+
+		ginkgo.By("Waiting for all PVCs to be Bound...")
+
+		select {
+		case <-waitForProvisionCh:
+			framework.Logf("All PVCs in Bound state")
+		case <-time.After(testTimeout):
+			ginkgo.Fail(fmt.Sprintf("expected all PVCs to be in Bound state within %v minutes", testTimeout))
+		}
+
+		ginkgo.By("Calculating performance metrics for provisioning operations")
+		createPerformanceStats(provisioningStats, l.pvcs)
+
+		ginkgo.By("Validating performance metrics against baseline")
+		errList := validatePerformanceStats(provisioningStats.operationMetrics)
+		framework.ExpectNoError(errors.NewAggregate(errList), "while validating performance metrics")
+
+	})
+
+}
+
+// createPerformanceStats calculates individual metrics for an operation
+// given the intervals collected during that operation
+func createPerformanceStats(stats *performanceStats, pvcs []*v1.PersistentVolumeClaim) {
+	var min, max, sum time.Duration
+	for _, pvc := range pvcs {
+		pvcMetric, ok := stats.perObjectInterval[pvc.Name]
+		framework.ExpectEqual(ok, true)
+
+		elapsedTime := pvcMetric.elapsed
+		sum += elapsedTime
+		if elapsedTime < min || min == 0 {
+			min = elapsedTime
+		}
+		if elapsedTime > max {
+			max = elapsedTime
+		}
+	}
+	stats.operationMetrics = &metrics{
+		latency:    time.Duration(int64(sum) / provisionCount),
+		throughput: provisionCount / max.Seconds(),
+	}
+}
+
+// validatePerformanceStats validates if test performance metrics meet the baseline target
+func validatePerformanceStats(operationMetrics *metrics) []error {
+	var errList []error
+	framework.Logf("Metrics to evaluate against baseline: %+v", spew.Sdump(operationMetrics))
+
+	if operationMetrics.latency > expectedLatency {
+		err := fmt.Errorf("expected latency to be less than %v but calculated latency %v", expectedLatency, operationMetrics.latency)
+		errList = append(errList, err)
+	}
+	if operationMetrics.throughput < expectedThroughput {
+		err := fmt.Errorf("expected throughput to be greater than %f but calculated throughput %f", expectedThroughput, operationMetrics.throughput)
+		errList = append(errList, err)
+	}
+	return errList
+}
+
+// newPVCWatch creates an informer to check whether all PVCs are Bound
+// When all PVCs are bound, the controller sends a signal to
+// waitForProvisionCh to unblock the test
+func newPVCWatch(f *framework.Framework, pvcMetrics *performanceStats) cache.Controller {
+	count := 0
+	countLock := &sync.Mutex{}
+	ns := f.Namespace.Name
+	checkPVCBound := func(oldPVC *v1.PersistentVolumeClaim, newPVC *v1.PersistentVolumeClaim) {
+		now := time.Now()
+		pvcMetrics.mutex.Lock()
+		defer pvcMetrics.mutex.Unlock()
+		countLock.Lock()
+		defer countLock.Unlock()
+
+		// Check if PVC entered the bound state
+		if oldPVC.Status.Phase != v1.ClaimBound && newPVC.Status.Phase == v1.ClaimBound {
+			newPVCInterval, ok := pvcMetrics.perObjectInterval[newPVC.Name]
+			framework.ExpectEqual(ok, true, "PVC %s should exist in interval map already", newPVC.Name)
```

Comment:
If we have an assertion in a go routine, we need to also call GinkgoRecover() so that errors are handled appropriately. See [example](https://github.com/kubernetes/kubernetes/blob/05a46dbb6026dfd9ae11600f8dc08a99994acdec/test/e2e/storage/testsuites/stress.go#L170)